### PR TITLE
Implement ICollection to DataTableWrapper

### PR DESCRIPTION
**Pull request by anderson-joyle**
For #637

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Core/Public/Values/CollectionTableValue.cs [hunk 1]
```diff
@@ -146,7 +146,7 @@ public override async Task<DValue<BooleanValue>> RemoveAsync(IEnumerable<Formula
 
                 foreach (var delete in deleteList)
                 {
-                    _sourceList.Remove(delete);
+                    ret = ret && _sourceList.Remove(delete);
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Values/CollectionTableValue.cs:149]: >= ret && [](http://example.com/codeflow?start=24&length=8)

nit:
&&=

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/PADIntegrationTests.cs [hunk 2]
```diff
@@ -121,6 +121,14 @@ public void DataTableEvalTest2()
         { Other : 5}
      )).Other");
             Assert.IsType<BlankValue>(result4);
+
+            var symbol = new SymbolTable();
+            symbol.EnableMutationFunctions();
+
+            engine.Config.SymbolTable = symbol;
+
+            var result5 = engine.Eval("Remove(robintable, {Names:\"name2\"});CountRows(robintable)", options: new ParserOptions() { AllowsSideEffects = true });
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/PADIntegrationTests.cs:130]: This is perfect!

2 additional asserts:
- before this eval, we should assert table.CountRows == 3. 
- after this eval, we should assert table.CountRows == 2 (ie, that the actual table object is modified)

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/PADIntegrationTests.cs [hunk 3]
```diff
@@ -121,6 +121,14 @@ public void DataTableEvalTest2()
...
+            Assert.Equal(2, ((NumberValue)result5).Value);
         }
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/PADIntegrationTests.cs:132]: >} [](http://example.com/codeflow?start=8&length=1)

Can you expand this to include calls to Patch() and Collect()? With similar asserts.

**Review by MikeStall — DISMISSED** [review]
:shipit:

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Review by MikeStall — APPROVED** [review]
:shipit:

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/PADIntegrationTests.cs [hunk 4]
```diff
@@ -121,6 +121,27 @@ public void DataTableEvalTest2()
...
             Assert.IsType<BlankValue>(result4);
+
+            var symbol = new SymbolTable();
+            var opt = new ParserOptions() { AllowsSideEffects = true };
+
+            symbol.EnableMutationFunctions();
+
+            engine.Config.SymbolTable = symbol;
+
+            Assert.Equal(3, table.Rows.Count);
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/PADIntegrationTests.cs:132]: >  Assert.Equal(3, table.Rows.Count); [](http://example.com/codeflow?start=10&length=36)

nit:
Assert.Equals("name1,name2,name3", Dump(table));


Assert.Equals("name1,name3", Dump(table));


Assert.Equals("name1,name3,name100", Dump(table));

Assert.Equals("new-name,name3,name100", Dump(table));

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/Scenarios/DataTableMarshallerProvider.cs [hunk 5]
```diff
@@ -63,6 +69,43 @@ protected override DValue<RecordValue> Marshal(DataRow item)
             return DValue<RecordValue>.Of(record);
         }
 
+        protected override DataRow MarshalInverse(RecordValue row)
+        {
+#pragma warning disable CA2000 // Dispose objects before losing scope
+            var dataTable = new DataTable();
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/Scenarios/DataTableMarshallerProvider.cs:75]: >var dataTable = new DataTable(); [](http://example.com/codeflow?start=12&length=32)

can we reuse existing _table?

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/Scenarios/DataTableMarshallerProvider.cs [hunk 6]
```diff
@@ -63,6 +69,43 @@ protected override DValue<RecordValue> Marshal(DataRow item)
...
+                dict[field.Name] = field.Value;
+            }
+
+            var index = 0;
+
+            foreach (DataColumn column in _table.Columns)
+            {
+                dataTable.Columns.Add(column.ColumnName, column.DataType);
+
+                if (dict.TryGetValue(column.ColumnName, out FormulaValue formulaValue))
```
[the remark this answers is not in the document]
Comment (anderson-joyle) [reply src/tests/Microsoft.PowerFx.Interpreter.Tests/Scenarios/DataTableMarshallerProvider.cs:92]: Method is protected and there is no "easy" way to get it reachable without changing extra places.

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/Scenarios/DataTableMarshallerProvider.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (anderson-joyle) [reply src/tests/Microsoft.PowerFx.Interpreter.Tests/Scenarios/DataTableMarshallerProvider.cs]: RecordValue.GetField()

**Review by MikeStall — APPROVED** [review]
:shipit:

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Pull request merged**